exploration/0003-formatted-parts.md

```diff
@@ -0,0 +1,187 @@
+
+## Proposed Design
+
+The formatted-parts API is included in the spec as an optional but recommended formatting target.
+
+The shape of the formatted-parts output is defined in a manner similar to the data model,
+which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.
+
+At the top level, the formatted-parts result is an iterable sequence of parts.
+Parts corresponding to each _text_ can be simpler than those of _expressions_,
+as they do not have a `source` other than their `value`,
+or set any of the other possible metadata fields.
+
+```ts
+type MessageParts = Iterable<
+  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
+>;
+
+interface MessageTextPart {
+  type: "text";
+  value: string;
+}
+```
+
+For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
+The parts' `dir` and `locale` may be defined if overridden by an expression attribute or set otherwise.
+Each part should have at most one of `value` or `parts` defined;
+some may have none.
+
+```ts
+interface MessageExpressionPart {
+  type: string;
+  source: string;
+  parts?: Iterable<{ type: string; value: unknown }>;
+  value?: unknown;
+  dir?: "ltr" | "rtl" | "auto";
+  locale?: string;
+}
+```
+
+The bidi isolation strategies included in the spec may require
+the insertion of MessageBiDiIsolationParts in the formatted-parts output.
+
+```ts
+interface MessageBiDiIsolationPart {
+  type: "bidiIsolation";
+  value: "\u2066" | "\u2067" | "\u2068" | "\u2069"; // LRI | RLI | FSI | PDI
+}
+```
+
+Some of the MessageExpressionPart instances may be further defined
+without reference to the function registry.
+
+Unannotated expressions with a _literal_ operand
+are represented by MessageStringPart.
+As with MessageTextPart,
+the `value` of MessageStringPart is always a string.
+
+```ts
+interface MessageStringPart {
+  type: "string";
+  source: string;
+  value: string;
+  dir?: "ltr" | "rtl" | "auto";
+  locale?: string;
+}
+```
+
+Unannotated expressions with a _variable_ operand
+that is not formatted according to its type
+are represented by MessageUnknownPart.
+
+```ts
+interface MessageUnknownPart {
+  type: "unknown";
+  source: string;
+  value: unknown;
+}
+```
+
+When the resolution or formatting of a placeholder fails,
+it is represented in the output by MessageFallbackPart.
+No `value` is provided; when formatting to a string,
+the part's representation would be `'{' + source + '}'`.
+
+```ts
+interface MessageFallbackPart {
+  type: "fallback";
+  source: string;
+}
+```
+
+Formatting functions defined in the registry
+will need accompanying formatted-parts representations.
+Where available, such a formatted value should itself be represented by `parts`
+rather than a unitary string `value`.
+These sub-parts should not need fields beyond their `type` and `value`,
+and in most cases it's presumed that the sub-part `value` would be a string.
+
+```ts
+interface MessageDateTimePart {
```

Comment:
Can we define the "parts" so that they are generic rather than each type having its own special part type?

Reply:
Yes, that's the MessageExpressionPart definition above, which these interfaces also match. The definitions here are giving more specificity about what e.g. `:datetime` and `:number` end up producing, i.e. that they have explicit `type` identifiers and define `parts` rather than `value`.